processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -331,90 +334,147 @@ public static void validateRowValues(
     final Indexed<String> dim1Names = adapter1.getDimensionNames();
     final Indexed<String> dim2Names = adapter2.getDimensionNames();
     for (int i = 0; i < dims1.length; ++i) {
-      final int[] dim1Vals = dims1[i];
-      final int[] dim2Vals = dims2[i];
+      final Object dim1Vals = dims1[i];
```

Comment:
I've moved the type-specific parts of this IndexIO validation section into a method on DimensionHandler